### DISCUSSION TITLE
Automatic redirects using hooks and the Git history

### DISCUSSION THREAD

**Discussion by kamilkrzyskow** [mkdocs/mkdocs#3134] [Show and tell]
> **Disclaimer:** This guide expects a basic understanding of the Python programming language, used in the solution. 
However, I provide a reusable script you can just use without full understanding. [This free course](https://cs50.harvard.edu/python) can also get you up to speed with using Python.

---

# Context
The [mkdocs-redirects](https://github.com/mkdocs/mkdocs-redirects) plugin allows to create _dynamic*_ redirections for older links.

```yaml
plugins:
    - redirects:
        redirect_maps:
            'old.md': 'new.md'
            'old/file.md': 'new/file.md'
            'some_file.md': 'https://external.url.com/foobar'
```

However, they're not fully dynamic, because the user has to maintain the `redirect_maps` manually after each location change. 
The user might forget to update them, or even make a mistake and write the old name with a typo or fat-finger and write a completely wrong destination.

Why not write the redirections automatically then to save time and headaches?

## Solution
Since version [1.4.0](https://github.com/mkdocs/mkdocs/releases/tag/1.4.0) MkDocs added support for [hooks](https://www.mkdocs.org/user-guide/configuration/#hooks). In short hooks are scripts that mimic the functionality of MkDocs plugins, but do not need to be packaged and installed to the Python environment. Having access to the plugin [events](https://www.mkdocs.org/dev-guide/plugins/#events) we can run operations during the build process.

Using the `GitPython` python module, it is possible to analyze and process the Git history. 
So it is possible to read commit data and their diff comparison, and detect files renames, which will work as base for the `redirect_maps` creation.

### Dependencies
`GitPython` is already required in plugins like [mkdocs-git-revision-date-localized-plugin](https://github.com/timvink/mkdocs-git-revision-date-localized-plugin), but if you're not using anything like that then install `GitPython` and/or include it into your `requirements.txt`:

```bash
pip install GitPython
```

of course don't forget about the `mkdocs-redirects` plugin:

```bash
pip install mkdocs-redirects
```

### Configuration

When deploying to GitHub Actions via workflow files, make sure you fetch the full Git history:

```yml
jobs:
  deploy:
    runs-on: ubuntu-latest
    steps:
      - name: Install Checkout
        uses: actions/checkout@v3
        with:
          # Commits to fetch. 0 indicates whole history.
          # Default: 1
          fetch-depth: 0
```

---

The hook assumes the parent directory of the docs as the repository root with the `.git` directory:
```
root_directory/
├─ .git/
├─ docs/
│  ├─ index.md
│  ├─ ...
├─ overrides/
│  ├─ .hooks/
│  │  ├─ git_redirects.py
├─ mkdocs.yml
```
> Created with [ASCII Generator](https://ascii-tree-generator.com/)

If that's not the case then you'll need to adjust this line: 
```py
project_root: Path = docs_dir.parent
```

---

Settings needed in the `mkdocs.yml` are:
```yml
hooks:
  - overrides/.hooks/git_redirects.py # Path may be different

# ...

plugins:
  - search
  # ...
  - redirects:
      redirect_maps: {} # You can add your own mapping too
```

--- 

You can limit the range of the processed commits by editing the constants:

```py
MAX_DATE: datetime.datetime = datetime.datetime.fromisoformat("2023-02-25")
"""The oldest included commit date in the history lookup."""

MAX_HASH: str = ""
"""The oldest included commit hash in the history lookup."""
```

`MAX_DATE` can be a range of days too:
```py
MAX_DATE: datetime.datetime = datetime.datetime.now() - datetime.timedelta(days=99)
```
It can, also be set as `None` to turn it off.

`MAX_HASH` is the commit revision hash, it can also be empty to turn it off.

### Methodology / How it works

The hook goes over the Git history and compares commits with their most recent parent. **It only accepts Rename diff items**, so when moving files around in the repository make sure they're marked as renamed in `git status` instead of `deleted/added`.
You should be able to move the file and fix relative paths in them. **Beware that a small file with a lot of paths might be marked as a complete deletion and addition if you decide to fix the paths in the same commit as the renaming.** 

```bash
$ git status
On branch main

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
        renamed:    docs/old.md -> docs/new_path/new.md
```

#### Limitations / opinionated quirks:
- Only `.md` -> `.md` file renames are accepted, so no support for `.md` -> `.md.off` -> `.md` etc.
- Only file renames inside the `docs` directory are accepted.
- Only files with 1 dot are accepted aka. `index.md` and so on. This blocks hidden files like `.index.md` or localized files `index.en.md`.
- If a new file already exists for a possible old redirection then the redirection won't be created.
- If the new destination doesn't exist then the redirection won't be created.
- If a file path is often changed renamed, only the most recent change will be kept as redirection.
- When a renaming chain occurs, all redirections will point to the newest name. 

### Example terminal display

```bash
$ mkdocs build -s
INFO     -  git_redirects: Processing Git history...
INFO     -  git_redirects: Created 48 redirects
INFO     -  Cleaning site directory
INFO     -  Building documentation to directory: site/
INFO     -  Documentation built in 4.95 seconds
```

---
 
The script hook works on both Windows and Linux (tested only on the GitHub Actions runner Ubuntu). 

Here is the version, this guide was basing on [git_redirects.zip](https://github.com/mkdocs/mkdocs/files/10869286/git_redirects.zip)
The newest version can also be [found here](https://github.com/Gothic-Modding-Community/gmc/blob/dev/overrides/.hooks/git_redirects.py).

Possible issues:

- When traversing the Git history from the beginning it might be quite slow for bigger repositories. My idea to solve it is to add caching, but it is on the back burner.
- There is no support for a toggle between LocalDev and Deployment builds, but it is on the back burner.
- I assume some people might want some more customization like getting only specific ranges of commit hashes etc. I'm not planning to add anything of that sort.

**Comment by squidfunk** [comment]
This sounds like an excellent case for a plugin, maybe something like `mkdocs-auto-redirects`? It would make it easier for users to use, because now they could just install the plugin and don't have to maintain a separate hook 😀

**Reply by kamilkrzyskow** [reply]
Sure, I agree that both of my hooks, this one and https://github.com/squidfunk/mkdocs-material/discussions/5060 would be great as plugins, because more people could use them then. Most people are used to configuring `yml` files with `mkdocs.yml` or GitHub workflow files and plugins don't gatekeep with the basic Python know-how disclaimer 😅. 

However it's not a priority at the moment, those hooks are a side-effect of figuring out stuff for the Gothic Modding Community repository, where I know too little to contribute via adding articles to the docs, so since I like Python I do play around with the _"backend"_ of our docs. Especially making everything compatible with the mkdocs-static-i18n plugin 😆 

https://github.com/timvink/mkdocs-git-revision-date-localized-plugin and similar plugins from Tim Vink use the `mkdocs-git-...` pattern, so I'm more inclined to use `mkdocs-git-redirects` 😄 

Thanks for the feedback!

**Comment by Andre601** [comment]
One issue I can see with this is, that git may not see a file move as such.

I had it more than once where moving a file to a new location and editing existing files in that location was considered a deletion of the old file and edit of the new, so no move of the file.

It's a rare case but still possible.

**Reply by kamilkrzyskow** [reply]
That's correct and is mentioned in the post above. There is some internal diff threshold until the file is considered a file rename, after that it's seen as a new file. Depends on the file size and the amount of changes to it, aka the ratio. Best to make the movement in one commit and then any link fixing in another to assure proper Git Blame preservation.

**Comment by Andre601** [comment]
One thing I wish the redirect plugin had was defining redirects in front matter.

Like...
```markdown
---
title: page
description: Some page

redirects:
  - old/location/index.md
  - another/old/location.md
---

Content.
```
It would have the benefit of being able to move pages, without having to update the redirect map, as the info is given by the page itself. It would also be more organized imo.

Obviously you still would need to add entries, but for me it feels easirr if I move the file and directly edit it, rather than go back to mkdocs.ym and edit that.

Perhaps I'll suggest this to the dev...

Also, I feel like that git-based redirect could fit as an option in the redirect plugin.

**Reply by kamilkrzyskow** [reply]
Here you go, https://github.com/kamilkrzyskow/gmc/commits/redirect-plugin?since=2023-02-24 that was my previous approach with [another hook](https://github.com/kamilkrzyskow/gmc/blob/redirect-plugin/overrides/.hooks/meta_redirects.py) and `old_nav` meta list of old paths. However, I decided against it because moving a whole directory of files into another location requires modifying all of those files, and this just sullies the Git Blame imo, and is rather tedious to do (trust me, I took it till the end and modified 40 files by hand).

The idea came from the solution here https://github.com/squidfunk/mkdocs-material/discussions/3758 the difference here, which is VERY important is that the ContextID is static, possibly managed by some outside tooling and database, so the files can move however they want and the redirection will always point to the correct path, this doesn't go well with manually readjusting the "ContextID" equivalent after each move.

**Reply by kamilkrzyskow** [reply]
Answered in your Issue in the `mkdocs-redirects` plugin. For further discussion let's move there https://github.com/mkdocs/mkdocs-redirects/issues/53
